### ISSUE TITLE
make service file for spis systemctl

### ISSUE THREAD

**Issue by id4vip** [gbbirkisson/spis#100]
I am very happy using spis on my machine with docker format. Now I want to install spis on debian machine directly. I have downloaded spis binary & install niginx & sqlite on my machine. Now I want that spis server run automatically after each time system reboot with systemctl command please guide me & help me to make service file for spis “ sudo nano /etc/systemd/system/spis.service” . I remembered I used for zigbee2mqtt this service file
`
[Unit]
Description=zigbee2mqtt
After=network.target

[Service]
Environment=NODE_ENV=production
ExecStart=/usr/bin/npm start
WorkingDirectory=/opt/zigbee2mqtt
StandardOutput=inherit
StandardError=inherit
Restart=always
RestartSec=10s
User=pi

[Install]
WantedBy=multi-user.target

`
Please modify this service file for spis software binary.
I save spis binary in  this location “/mnt/mydata/spis/spis-server-x86_64-unknown-linux-gnu”
I want spis dedicated data folder “/mnt/mydata/spis_data/”
My images are at “/mnt/mydata/my_photos” want photos are read only permission for spis
Please tell me how I can pass these location values to spis & binary  software control by ststemctl. thanks

**Comment by id4vip** [comment]
Thanks for helping, I have made my spis.service file 
`
[Unit]                                                                                          
Requires=multi-user.target                                                                      
After=multi-user.target                                                                         
                                                                                                
[Service]                                                                                       
Environment="SPIS_MEDIA_DIR=/mnt/dietpi_userdata/plex_data"                                     
Environment="SPIS_DATA_DIR=/mnt/dietpi_userdata/spis/data"                                      
Environment="SPIS_SERVER_SOCKET=/var/run/spis.sock"                                             
Environment="SPIS_API_MEDIA_PATH=/mnt/dietpi_userdata/spis/media"                               
Environment="SPIS_API_THUMBNAIL_PATH=/mnt/dietpi_userdata/spis/data/thumbnails"                 
Environment="RUST_LOG=error,spis_server=info"                                                   
User=spis                                                                                       
Group=spis                                                                                      
ExecStart=/mnt/dietpi_userdata/spis/spis_active                                                 
Restart=always                                                                                  
RestartSec=5s                                                                                   
                                                                                                
[Install]                                                                                       
WantedBy=multi-user.target  
`
but system fail to start program. the logs are like this

`
Jul 31 19:10:12 lxc501 systemd[1]: Started spis.service.
Jul 31 19:10:12 lxc501 spis_active[8645]: 2023-07-31T18:10:12.637087Z  INFO spis_server: Starting spis version 0.5.3
Jul 31 19:10:12 lxc501 spis_active[8645]: 2023-07-31T18:10:12.637513Z  INFO spis_server: Media dir: "/mnt/dietpi_userdata/plex_data"
Jul 31 19:10:12 lxc501 spis_active[8645]: 2023-07-31T18:10:12.637566Z  INFO spis_server: Thumb dir: "/mnt/dietpi_userdata/spis/data/thumbnails"
Jul 31 19:10:12 lxc501 spis_active[8645]: 2023-07-31T18:10:12.637609Z  INFO spis_server: Data file: "/mnt/dietpi_userdata/spis/data/spis.db"
Jul 31 19:10:12 lxc501 spis_active[8645]: 2023-07-31T18:10:12.637648Z  INFO spis_server: Setting up DB
Jul 31 19:10:12 lxc501 spis_active[8645]: 2023-07-31T18:10:12.638599Z  INFO spis_server: Setting up media processing
Jul 31 19:10:12 lxc501 spis_active[8645]: 2023-07-31T18:10:12.638718Z  INFO spis_server: Setting up file watcher
Jul 31 19:10:12 lxc501 spis_active[8645]: 2023-07-31T18:10:12.638942Z  INFO spis_server: Setting up file walker
Jul 31 19:10:12 lxc501 spis_active[8645]: 2023-07-31T18:10:12.639013Z  INFO spis_server: Start listening on socket /var/run/spis.sock
Jul 31 19:10:12 lxc501 spis_active[8645]: 2023-07-31T18:10:12.639101Z  INFO spis_server::pipeline: Starting db processing component
Jul 31 19:10:12 lxc501 spis_active[8645]: 2023-07-31T18:10:12.639145Z  INFO spis_server::pipeline: Added processing schedule: 0 0 2 * * *
Jul 31 19:10:12 lxc501 spis_active[8645]: The application panicked (crashed).
Jul 31 19:10:12 lxc501 spis_active[8645]: Message:  Failed to create server:
Jul 31 19:10:12 lxc501 spis_active[8645]:    0: Permission denied (os error 13)
Jul 31 19:10:12 lxc501 spis_active[8645]: Location:
Jul 31 19:10:12 lxc501 spis_active[8645]:    spis-server/src/server/mod.rs:171
Jul 31 19:10:12 lxc501 spis_active[8645]: Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Jul 31 19:10:12 lxc501 spis_active[8645]: Run with RUST_BACKTRACE=full to include source snippets.
Jul 31 19:10:12 lxc501 spis_active[8645]: Location: spis-server/src/main.rs:120
Jul 31 19:10:12 lxc501 spis_active[8645]: Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Jul 31 19:10:12 lxc501 spis_active[8645]: Run with RUST_BACKTRACE=full to include source snippets.
Jul 31 19:10:12 lxc501 spis_active[8645]: 2023-07-31T18:10:12.641028Z  WARN spis_server::pipeline: file_channel was closed
Jul 31 19:10:12 lxc501 systemd[1]: spis.service: Main process exited, code=exited, status=101/n/a
Jul 31 19:10:12 lxc501 systemd[1]: spis.service: Failed with result 'exit-code'.
`

**Comment by gbbirkisson** [comment]
TL;DR: The user `spis` is not allowed to create the socket `/var/run/spis.sock` on your system.

I guess the error handling could be better here, but the logs state that the error originates from `spis-server/src/server/mod.rs:171`. So looking at that snippet ...

https://github.com/gbbirkisson/spis/blob/50ff25bc6a95ed3dcf098f35dfbcc0618da0cff4/spis-server/src/server/mod.rs#L165-L174

... we can see that it fails to create the socket.

**Comment by id4vip** [comment]
Thanks, for your tip, after adding sock file with permission. software start working. I configured nginx also so now able to see spis default page with time line on top. when I have added few photo to test. system not able to shows those on web page. in logs it is complaining exif data not found. logs are attached with link. just for clarify the photo files need to be spis user permission/domination or i can give photos to read only access to spis software by giving photos under plex user with read only access to spis. i do not want anybody delete photos through spis webpage. 
https://dpaste.org/28fEC
also the photo without exif data like scanned etc. will not able to shown through spis system. I want all photos must be shown even time line screwed.

**Comment by gbbirkisson** [comment]
> I want all photos must be shown even time line screwed.

Creating a separate issue for that: https://github.com/gbbirkisson/spis/issues/101

> and in nginx my reverse proxy conf is this

Use the nginx config file from the dockerfile as a reference to create your new nginx config:

https://github.com/gbbirkisson/spis/blob/50ff25bc6a95ed3dcf098f35dfbcc0618da0cff4/docker/nginx.conf#L1-L35

**Comment by id4vip** [comment]
The reason for using 0.0.0.0:8085 port/web based ip address is that. I want to avoid nginx & want to use rust base reverse proxy
https://github.com/junkurihara/rust-rpxy/tree/develop
this proxy this time working at my home and dealing with home-assistant, vaultwarden, adguard-home & plex. is there any limitation using port based approach as i do not know how i can connect spis with rpxy through web socket etc.

**Comment by gbbirkisson** [comment]
> is there any limitation using port based approach

Nope, its just a matter of preference

**Comment by gbbirkisson** [comment]
You have told SPIS that the proxy will serve the images on these paths:

```
Environment="SPIS_API_MEDIA_PATH=/mnt/dietpi_userdata/spis_data/media"
Environment="SPIS_API_THUMBNAIL_PATH=/mnt/dietpi_userdata/spis_data/thumbnails"
```

That translates to:

```
http://spis.XXX.ZZZ/mnt/dietpi_userdata/spis_data/media/someimage.jpg
http://spis.XXX.ZZZ/mnt/dietpi_userdata/spis_data/thumbnails/somethumbnail.jpg
```

I suspect that this is incorrect. Use the browser network tab to see where its trying to get the images from. Make sure that your reverse proxy is serving the images at those locations, or fix the configuration and point SPIS to the correct paths.

**Comment by id4vip** [comment]
Thanks, now i am understand the trouble. your project is perfect but my bad luck i am not able to bridge two rust projects. in other projects only one page is divert by reverse proxy & all other functioning carried by 3rd party software. in spis we need to give default page & thumbnail address & photo address to revers proxy as your nginx.conf. it will be work with nginx but i do not know how to mould in rpxy (https://github.com/junkurihara/rust-rpxy) . i am getting this error in rpxy logs with not finding thumbnails. I am not much expert in linux so not understanding what is the remedy to bridge rpxy with spis (both awesome project in rust) thanks

**Comment by id4vip** [comment]
my log file in rpxy side is like
`
Aug 02 17:34:28 lxc501 rpxy_active[9419]: 2023-08-02T16:34:28.287643Z  INFO rpxy rpxy_lib::log: spis.XXX.ZZZ:8080 <- 192.168.1.93:64300 -- GET /mnt/dietpi_userdata/spis_data/thumbnails/7d660794-6e1a-3c26-b5b8-87dc65f966b0.webp HTTP/1.1 -- 404 Not Found --  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36", "192.168.1.93" "http://localhost:8085/mnt/dietpi_userdata/spis_data/thumbnails/7d660794-6e1a-3c26-b5b8-87dc65f966b0.webp"

`
I have checked in given folder spis making thumbnail. but I am not manipulate that in rpxy revers proxy side.

**Comment by gbbirkisson** [comment]
Opening the port SPIS is running on (`http://192.168.1.150:8085/`) will serve you the GUI, but not the images. This is stated in the README!

> Note that the spis-server binary does not serve images. For that you can use something like nginx. See [nginx config](https://github.com/gbbirkisson/spis/blob/main/docker/nginx.conf) for an example.

You have to serve both SPIS and static images through your proxy:

```mermaid
flowchart LR

A[Browser] -->|Request on 8080| B[Proxy]
B -->|Request on 8085| C[SPIS]
B -->|Serve Images| D[Files on disk]
```

Again, read the nginx config provided in this repo, understand how it works, and configure your own proxy accordingly:

https://github.com/gbbirkisson/spis/blob/50ff25bc6a95ed3dcf098f35dfbcc0618da0cff4/docker/nginx.conf#L1-L35

**Comment by id4vip** [comment]
thanks for helping me. rust-rpxy is not compatible with SPIS software by it self , it need nginx backend support from nginx for web file sharing.

**Comment by id4vip** [comment]
my exact spis.service file is 
`
[Unit]
Requires=multi-user.target
After=multi-user.target

[Service]
Environment="SPIS_MEDIA_DIR=/mnt/dietpi_userdata/plex_data"
Environment="SPIS_DATA_DIR=/mnt/dietpi_userdata/spis_data"
Environment="SPIS_SERVER_SOCKET=/mnt/dietpi_userdata/spis/spis.sock"
Environment="SPIS_API_MEDIA_PATH=/mnt/dietpi_userdata/spis_data/media"
Environment="SPIS_API_THUMBNAIL_PATH=/mnt/dietpi_userdata/spis_data/thumbnails"
Environment="RUST_LOG=error,spis_server=info"
Environment="SPIS_SERVER_ADDRESS=0.0.0.0:8085"
Environment="SPIS_PROCESSING_RUN_ON_START=true"
Environment="SPIS_PROCESSING_SCHEDULE=0 0 22 * * *"
User=spis
Group=spis
ExecStart=/mnt/dietpi_userdata/spis/spis_active
Restart=always
RestartSec=5s

[Install]
WantedBy=multi-user.target

`

& for nginx my spis.conf is
`
server {
#    listen       ${NGINX_PORT};
#    listen  [::]:${NGINX_PORT};
    listen       80;
    listen  [::]:80;
    server_name spis.zzz.xxx;


    access_log /dev/stdout;

    gzip on;
    gzip_proxied any;
    gzip_types
        text/html
        application/wasm
        application/javascript
        application/json
        text/plain
        image/jpeg;

    location /mnt/dietpi_userdata/spis_data/media {
        gzip_static on;
        expires 1y;
        add_header Cache-Control "public";
        alias /mnt/dietpi_userdata/plex_data;
    }

    location /mnt/dietpi_userdata/spis_data/thumbnails {
        gzip_static on;
        expires 1y;
        add_header Cache-Control "public";
        alias /mnt/dietpi_userdata/spis_data/thumbnails;
    }

    location / {
        proxy_pass http://unix:/mnt/dietpi_userdata/spis/spis.sock;
        proxy_http_version 1.1;
    }
}

`

my spis software is working. but nginx is failing to start , the log of nginx is as.
https://dpaste.org/Oxn7h
Please guide me how can access spis whole system from my network through nginx revers proxy access my this address "http://spis.zzz.xxx" . my adgurdhome diverting all *.zzz.xxx to my local access 192.168.1.75 . if i disable spis.conf file in nginx then nginx works and able to access homeassistant, adguardhome plex etc. only there is trouble in spis.conf. 

the nginx command "nginx -t" give this result
`
root@lxc501:/etc/nginx/sites-enabled# nginx -t
2023/08/07 22:22:02 [warn] 3574#3574: duplicate MIME type "text/html" in /etc/nginx/sites-enabled/spis.conf:19
nginx: the configuration file /etc/nginx/nginx.conf syntax is ok
nginx: configuration file /etc/nginx/nginx.conf test is successful

`

**Comment by id4vip** [comment]
one more clarification question, what is the relation ship between "SPIS_API_MEDIA_PATH" variable & "SPIS_MEDIA_DIR" .
as i declare "SPIS_MEDIA_DIR" is my plex_data. i want spis just read that data but for all his work i gave different folder which software owned itself. is this approach is correct or these values must be same.

**Comment by id4vip** [comment]
I confirm after hash out "# access_log /dev/stdout;"  nginx & spis both service working. 
i am able to open thumbnail directly via to known file name
`
http://spis.zzz.xxx/mnt/dietpi_userdata/spis_data/thumbnails/2adb0b13-964f-3130-acc3-1d80945203ec.webp
`  
so confirm webserver is working. but  spis default page is not opening and giving me nginix 502 error. so something wrong in websocket adding format
`
location / {
    proxy_pass http://unix:/mnt/dietpi_userdata/spis/spis.sock;
    proxy_http_version 1.1;
}
`
please guide me for correct formatting for adding socket connection for nginx. overall both services working.

**Comment by id4vip** [comment]
just want to confirm ipaddress & port method works in nginx & websocket is not working. please guide me toward the correct formatting of websocket approach. I tried to remold author nginx file into my required folder & websocket format is not working. so something in websocket formatting is wrong. overall with ip:port approach spis functionality works
`
location / {
#    proxy_pass http://unix:/mnt/dietpi_userdata/spis/spis.sock;
    proxy_pass http://192.168.1.150:8085;
    proxy_http_version 1.1;
}
`

**Comment by gbbirkisson** [comment]
> Please guide me how can access spis whole system from my network ...

I'm afraid you will have to figure that out on your own.

> what is the relation ship between "SPIS_API_MEDIA_PATH" variable & "SPIS_MEDIA_DIR"  ...

- `SPIS_MEDIA_DIR` is what directory SPIS expects to find your files
- `SPIS_API_MEDIA_PATH` is what path your reverse proxy serves your files on

So a valid configuration could be:

```
# spis.service
...
Environment="SPIS_MEDIA_DIR=/mnt/dietpi_userdata/plex_data"
Environment="SPIS_API_MEDIA_PATH=/api/media"
...
```

```
# nginx.conf
...
     location /api/media { 
         gzip_static on; 
         expires 1y; 
         add_header Cache-Control "public"; 
         alias /mnt/dietpi_userdata/plex_data; 
     }
...
```

> as my spis service is working & not able to find any sock file in given folder location

You are specifically telling SPIS to listen to a port not a socket. Change it so SPIS listens to a socket:

```diff
# spis.service
...
-Environment="SPIS_SERVER_ADDRESS=0.0.0.0:8085"
+Environment="SPIS_SERVER_SOCKET=/mnt/dietpi_userdata/spis/spis.sock"
...
```

**Comment by id4vip** [comment]
this thing correct & works  proxy_pass http://unix:/mnt/my_userdata/myapp/myapp.sock:/; 
find the solution sock file need to be  chmod 777 /mnt/my_userdata/myapp/myapp.sock only then nginx open the webpage

**Comment by gbbirkisson** [comment]
Run SPIS and nginx with the same user/group and you should not have that problem
